src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java

```diff
@@ -13,12 +34,368 @@
+                continue;
+            }
+
+            boolean needsItem = false;
+            if(building instanceof AbstractBuildingWorker && ((AbstractBuildingWorker) building).neededForWorker(stack))
+            {
+                needsItem = true;
+            }
+            //todo count how much of the itemStack has to be left behind.
+            for(Map.Entry<ItemStorage, Integer> entry: building.needXForWorker().entrySet())
+            {
+                if(entry.getKey().getItem() == stack.getItem() && entry.getKey().getDamageValue() == stack.getItemDamage())
+                {
+                    needsItem = true;
+                }
+            }
+
+            if(!needsItem)
+            {
+                if(!worker.getInventoryCitizen().addItemStackToInventory(stack))
+                {
+                    return;
+                }
+                building.getTileEntity().removeStackFromSlot(i);
+            }
+        }
+    }
+
+    /**
+     * Gets a random building from his colony.
+     *
+     * @return a random blockPos.
+     */
+    @Nullable
+    private BlockPos getRandomBuilding()
+    {
+        if (worker.getColony() == null || getOwnBuilding() == null)
+        {
+            return null;
+        }
+
+        final Collection<AbstractBuilding> buildingList = worker.getColony().getBuildings().values();
+        final Object[] buildingArray = buildingList.toArray();
+
+        final int random = worker.getRandom().nextInt(buildingArray.length);
+        final AbstractBuilding building = (AbstractBuilding) buildingArray[random];
+
+        if (building instanceof BuildingWareHouse
+                || building instanceof BuildingDeliveryman || building instanceof BuildingTownHall)
+        {
+            return null;
+        }
+
+        return building.getLocation();
+    }
+
+    /**
+     * Dump the inventory into the warehouse.
+     * @return the next state to go to.
+     */
+    public AIState dump()
+    {
+        if(!worker.isWorkerAtSiteWithMove(wareHouse.getLocation(), MIN_DISTANCE_TO_WAREHOUSE))
+        {
+            return DUMPING;
+        }
+
+        wareHouse.getTileEntity().dumpInventoryIntoWareHouse(worker.getInventoryCitizen());
+        gatherTarget = null;
+
+        return START_WORKING;
+    }
+
+
+    /**
+     * Deliver the items to the hut.
+     * @return the next state.
+     */
+    private AIState deliver()
+    {
+        AbstractBuildingWorker ownBuilding = getOwnBuilding();
+        if(ownBuilding instanceof BuildingDeliveryman)
+        {
+            AbstractBuilding buildingToDeliver = ((BuildingDeliveryman) ownBuilding).getBuildingToDeliver();
+            if (buildingToDeliver != null)
+            {
+                if(!worker.isWorkerAtSiteWithMove(buildingToDeliver.getLocation(), MIN_DISTANCE_TO_WAREHOUSE))
+                {
+                    return DELIVERY;
+                }
+
+                InventoryCitizen workerInventory  = worker.getInventoryCitizen();
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'workerInventory' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)